### PR TITLE
Add CPPUnit Testing Framework

### PR DESCRIPTION
**Pull request by pbauman**
I've changed my mind about CPPUnit and am very much desiring the ease-of-adding new tests to the testing suite, especially in light of #163 etc. where we'll need to add more coverage.

I'd like to eventually migrate all unit tests (and regression tests?) to this framework. For now, I'm just trying to lay the groundwork to make adding new tests easy.

This is not complete yet for this PR, but the structure is there (using just the Arrhenius rate test) so that we can have a discussion about the setup.

My thinking was as follows:
* There are "standard" tests that don't require any external dependencies (modulo CPPUnit...). There's a program `standard_unit_tests` that will run all these.
* Each of the external dependencies (e.g. Eigen, VexCL, etc.) will have their own program to run. Currently, this PR has the `eigen_unit_tests`.

I've tried to style the tests to be as reusable as I can, but I'm sure there are parts I could do better. E.g. in the Eigen tests, I reuse the `ArrheniusRateVectorTestBase` class. So it's header only and then we just compile the sources needed for each in their own directory, e.g. `standard_unit/arrhenius_rate_valarray_test.C` and `eigen_unit/arrhenius_rate_eigen_test.C`.

The main reason I went this route is that it makes it clear what tests we're skipping and what we're not. This is as opposed to ifdef'ing out the tests and only have one-cppunit-program-to-rule-them-all. Thoughts on this setup?

If folks are OK with this setup, I'll drop the `MetaPhysicL`, `VexCL`, and `GSL` ones in and consider this PR complete for laying the ground work. Then we can migrate the existing tests in our spare time and add new ones to the CPPUnit framework.

### CODE REVIEW

**Comment by pbauman** [comment]
Discussion related to separating tests for dependencies happened offline with @roystgnr concluding
> Yeah, that's a fine idea.

So I will proceed with this.

File: test/standard_unit/arrhenius_rate_vector_test_base.h [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#include "antioch/physical_constants.h"
+#include "antioch/units.h"
+
+// Base class
+#include "arrhenius_rate_test_helper.h"
+#include "reaction_rate_vector_test_base.h"
+
+template<typename PairScalars>
+class ArrheniusRateVectorTestBase : public ArrheniusRateTestHelper</*Scalar*/typename Antioch::value_type<PairScalars>::type>,
+                                    public ReactionRateVectorBaseTest<Antioch::ArrheniusRate</*Scalar*/typename Antioch::value_type<PairScalars>::type>,PairScalars>
```
Comment (dmcdougall) [inline test/standard_unit/arrhenius_rate_vector_test_base.h]: Do you have a class naming inconsistency?  `*TestBase` versus `*BaseTest`.

File: test/standard_unit/arrhenius_rate_vector_test_base.h [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+// Antioch
+#include "antioch/physical_constants.h"
+#include "antioch/units.h"
+
+// Base class
+#include "arrhenius_rate_test_helper.h"
+#include "reaction_rate_vector_test_base.h"
+
+template<typename PairScalars>
+class ArrheniusRateVectorTestBase : public ArrheniusRateTestHelper</*Scalar*/typename Antioch::value_type<PairScalars>::type>,
```
Comment (dmcdougall) [inline test/standard_unit/arrhenius_rate_vector_test_base.h]: It's not clear that `ArreniusRateTestHelper` is actually needed.  Is there a reason against putting all the members of `ArrheniusRateTestHelper` in `ArrheniusRateVectorTestBase`?

**Comment by dmcdougall** [comment]
Can we enable doxygen for the test classes?  It'd be easier to see the class hierarchy that way.

File: test/standard_unit/arrhenius_rate_vector_test_base.h (hunk 2 above)
Comment (pbauman) [inline test/standard_unit/arrhenius_rate_vector_test_base.h]: I reuse `ArreniusRateTestHelper` in defining the non-vector tests [here](https://github.com/pbauman/antioch/blob/cpp-unit/test/standard_unit/arrhenius_rate_test.C#L44). So I think I'll leave it as is.

**Comment by pbauman** [comment]
All the optional library CppUnit structure is there now. In the last two commits, I turned on all the optional libraries when Doxygen gets built and then put all the testing in a separate namespace (`AntiochTesting`). I was hoping the latter might help organize the documentation better, but it didn't help all that much, IMHO. 

Thoughts on that? I was thinking maybe an alternative would be to opt-in at configure time to have the testing documentation built and by default it's not.

Regardless I've addressed the other comments (thanks @dmcdougall) and the testing infrastructure is there so this is ready to go barring comments on the documentation (and anything else that comes up if @roystgnr takes a look).

**Comment by dmcdougall** [comment]
> I turned on all the optional libraries when Doxygen gets built and then put all the testing in a separate namespace (AntiochTesting). I was hoping the latter might help organize the documentation better, but it didn't help all that much, IMHO.

There's a 'namespace' list that Doxygen generates so you can look at just the classes in that namespace.

**Comment by dmcdougall** [comment]
Looks good to me.  Good work @pbauman.

**Pull request merged**